### PR TITLE
[icu] Update, enable uwp, fix osx-dynamic

### LINKED ISSUE #33440: [icu] Build error
Package: icu:x64-osx-dynamic -> 73.1

**Host Environment**

- Host: x64-osx
- Compiler: AppleClang 14.0.3.14030022
-    vcpkg-tool version: 2023-08-09-9990a4c9026811a312cb2af78bf77f3d9d288416
    vcpkg-scripts version: 436a061a5 2023-08-25 (4 days ago)

**To Reproduce**

`vcpkg install --allow-unsupported`
**Failure logs**

```
-- Using cached icu4c-73_1-src.tgz.
-- Cleaning sources at phoenix/.cache/vcpkg/buildtrees/icu/src/c-73_1-src-3575e21c7c.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source phoenix/.cache/vcpkg/downloads/icu4c-73_1-src.tgz
-- Applying patch disable-escapestr-tool.patch
-- Applying patch remove-MD-from-configure.patch
-- Applying patch fix_parallel_build_on_windows.patch
-- Applying patch fix-extra.patch
-- Applying patch mingw-dll-install.patch
-- Applying patch disable-static-prefix.patch
-- Applying patch fix-win-build.patch
-- Using source at phoenix/.cache/vcpkg/buildtrees/icu/src/c-73_1-src-3575e21c7c.clean
-- Found external ninja('1.11.1').
-- Getting CMake variables for x64-osx-dynamic-dbg
-- Getting CMake variables for x64-osx-dynamic-rel
-- Generating configure for x64-osx-dynamic
-- Finished generating configure for x64-osx-dynamic
-- Configuring x64-osx-dynamic-dbg
-- Configuring x64-osx-dynamic-rel
-- Building x64-osx-dynamic-dbg
-- Building x64-osx-dynamic-rel
-- setting rpath prefix for macOS dynamic libraries
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: /usr/bin/install_name_tool -id @rpath/libicudata.73.dylib libicudata.73.1.dylib
    Working Directory: phoenix/.cache/vcpkg/buildtrees/icu/x64-osx-dynamic-dbg/lib
    See logs for more information:
      phoenix/.cache/vcpkg/buildtrees/icu/make-build-fix-rpath-x64-osx-dynamic-dbg-err.log

Call Stack (most recent call first):
  buildtrees/versioning_/versions/icu/acc3cf9137af5fc8fdd3e8607377026cf88f144d/portfile.cmake:95 (vcpkg_execute_build_process)
  scripts/ports.cmake:147 (include)



```
<details><summary>phoenix/.cache/vcpkg/buildtrees/icu/make-build-fix-rpath-x64-osx-dynamic-dbg-err.log</summary>

```
error: /Library/Developer/CommandLineTools/usr/bin/install_name_tool: can't open file: libicudata.73.1.dylib (No such file or directory)
```
</details>**Additional context**

<details><summary>vcpkg.json</summary>

```
{
  "name": "phoenix",
  "version-string": "1.0.0",
  "builtin-baseline": "4a8d11802f0a9ffbe5c68f6143d61d469d800a7d",
  "dependencies": [
    {
      "name": "gtest",
      "version>=": "1.13.0"
    },
    {
      "name": "opencv",
      "version>=": "4.8.0",
      "features": [
        "ade",
        "contrib",
        "cuda",
        "cudnn",
        "dc1394",
        "default-features",
        "dnn",
        "dnn",
        "eigen",
        "ffmpeg",
        "freetype",
        "gdcm",
        "gstreamer",
        "gtk",
        "halide",
        "ipp",
        "jasper",
        "jpeg",
        "lapack",
        "nonfree",
        "openexr",
        "opengl",
        "openmp",
        "ovis",
        "png",
        "python",
        "qt",
        "quirc",
        "sfm",
        "tbb",
        "tiff",
        "webp",
        "world"
      ]
    },
    {
      "name": "nlohmann-json",
      "version>=": "3.11.2"
    }
  ]
}

```
</details>

### PR DESCRIPTION
**Pull request by dg0yt**
Update to 74.1.
Simplify portfile.
Enable uwp.
Fix https://github.com/microsoft/vcpkg/issues/33440, osx-dynamic, and simplify the portfile by achieving the effect already during linking. As before, users can choose an alternative to `@rpath` via a triplet variable. In addition, they could complete disable the effect by supplying `--disable-rpath` via `CMAKE_CONFIGURE_MAKE_OPTIONS`.

### CODE REVIEW

**Comment by dg0yt** [comment]
Microsoft, do you accept this uwp hack:
https://github.com/unicode-org/icu/blob/da83309900e90509d6d559e4ad5f9a2a0a1cef6b/icu4c/source/common/umapfile.cpp#L41-L58
?

File: ports/icu/portfile.cmake [hunk 1]
```diff
@@ -20,6 +20,7 @@ vcpkg_extract_source_archive(SOURCE_PATH
         disable-static-prefix.patch # https://gitlab.kitware.com/cmake/cmake/-/issues/16617; also mingw.
         fix-win-build.patch
         vcpkg-cross-data.patch
+        darwin-rpath.patch
```
Comment (dg0yt) [inline ports/icu/portfile.cmake:23]: Related: https://github.com/unicode-org/icu/pull/2467, https://unicode-org.atlassian.net/browse/ICU-21458.

File: ports/icu/portfile.cmake [hunk 2]
```diff
@@ -35,6 +36,10 @@ elseif(VCPKG_TARGET_IS_UWP)
     vcpkg_list(APPEND CONFIGURE_OPTIONS --disable-extras ac_cv_func_tzset=no ac_cv_func__tzset=no)
     string(APPEND VCPKG_C_FLAGS " -DU_PLATFORM_HAS_WINUWP_API=1")
     string(APPEND VCPKG_CXX_FLAGS " -DU_PLATFORM_HAS_WINUWP_API=1")
+elseif(VCPKG_TARGET_IS_OSX)
+    if(DEFINED CMAKE_INSTALL_NAME_DIR)
+        vcpkg_list(APPEND BUILD_OPTIONS "ID_PREFIX=${CMAKE_INSTALL_NAME_DIR}")
+    endif()
```
Comment (dg0yt) [inline ports/icu/portfile.cmake:42]: Legacy port behavior.

**Comment by dg0yt** [comment]
icu is ready for review here.  
cppcms is ready for review in https://github.com/Microsoft/vcpkg/pull/35578, merged here to disable it for uwp.  
libzim is ready for review in https://github.com/Microsoft/vcpkg/pull/35576, merged here to disable it for uwp.

**Review by Adela0814 — APPROVED** [review]

**Comment by vicroms** [comment]
> Microsoft, do you accept this uwp hack: https://github.com/unicode-org/icu/blob/da83309900e90509d6d559e4ad5f9a2a0a1cef6b/icu4c/source/common/umapfile.cpp#L41-L58 ?

I'll ask the rest of the team about their opinion on this patch, otherwise the changes LGTM

**Comment by BillyONeal** [comment]
> > Microsoft, do you accept this uwp hack: https://github.com/unicode-org/icu/blob/da83309900e90509d6d559e4ad5f9a2a0a1cef6b/icu4c/source/common/umapfile.cpp#L41-L58 ?
> 
> I'll ask the rest of the team about their opinion on this patch, otherwise the changes LGTM

LGTM: we aren't the WACK

**Pull request merged**